### ISSUE TITLE
filepath.FilePath.remove breaks on symlinks

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3097]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)| @cyli reported|
|-|-|
|Trac ID|trac#3097|
|Type|enhancement|
|Created|2008-03-19 01:49:43Z|
|Branch|https://github.com/twisted/twisted/tree/filepath-remove-symlink-3097|

FilePath.remove on a symlink whose target is a directory will result in an OSError.  FilePath.remove currently checks to see if the path is a directory, and if so, removes all children and then calls os.rmdir on itself.  Otherwise, it calls os.remove on itself.  The problem is that a link that points to a directory returns true for both islink() and isdir().  Therefore, when you call FilePath.remove on a symlink whose target is a directory, it will both empty that target directory (which probably is not a desirable outcome) and call rmdir on the link, which raises an error.

Attachments:

* [filepath.remove.symlink.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/df4/df4f707507607e86a1f1ca735618fc6d06709986/05fe8102bcb0add778cd453df24a5cd01132e903.patch) (1125 bytes) - added by cyli on 2008-03-19 06:01:11Z - patch
* [filepath.remove.symlink.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/df4/df4f707507607e86a1f1ca735618fc6d06709986/89a2118396a55f70a3bb469ba39b1ec9328d289e.patch) (1690 bytes) - added by cyli on 2008-03-21 14:08:56Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3097 3097
type__enhancement enhancement
reporter__cyli cyli
priority__highest highest
milestone__ 
branch__branches_filepath_remove_symlink_3097 branches/filepath-remove-symlink-3097
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1205891383000000 1205891383000000
changetime__1206320060000000 1206320060000000
version__None None
owner__ 
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_1"></a>|@cyli removed owner|
|-|-|

This solves the problem.  I think this is the desired behavior for filepath.FilePath.remove

Another behavior that may be desirable (I didn't think so, so it's not in the patch) is that when filepath.FilePath.remove is called on a symlink that points to a directory: 1) everything inside the target directory is removed, 2) the target directory is removed,  and 3) the symlink to that directory is removed.  If this is the desired behavior, filepath.FilePath.remove can take an optional argument (ex. followLink) that defaults to False (since I don't think it's usually the desired behavior) but can be True when this behavior is desired.

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|spiv commented|
|-|-|

It's a shame that `remove` is implemented using look-before-you-leap, which is inherently susceptible to race conditions.  Although I guess any attempt to implement "delete this path, whatever sort of thing it is" is going to be racy on POSIX (because there's no atomic "`unlink` or `rmdir`" operation, let alone one that recurses), so this perhaps doesn't matter much in practice...

Getting more off-topic, having remove always mean recursively delete seems dubious to me.  There certainly ought to be a convenience method with those semantics, but I'd also like there to be a more primitive "remove a file or error" method, like `os.unlink`.  Maybe I'm just too used to thinking in terms of POSIX semantics :)

(These remarks should not be taken as criticism of the proposed fix, btw.)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve set owner to @cyli|
|-|-|

That looks sensible. My comment on the patch:
 * the new behavior should be tested in a new (different) test, not in testRemove. The lack of os.symlink should result to an explicit skip
 * it would be great if remove could gain a doctring in the process. I won't press you to do it though :).

Thanks.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_4"></a>|@cyli removed owner|
|-|-|

Docstring put in, and tests moved to a new test that explicitly skips if symlinks are not supported.  I won't change it here, since it's not part of the bug, but I could make remove take an optional argument as to whether or not to recursively remove directories.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [23048]) Branching to 'filepath-remove-symlink-3097'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [23050]) simple non-behavioral changes

use getattr instead of hasattr - hasattr can mask exceptions (probably not on a module, but who knows)
write docstring in a newer style than is prevalant in test_paths.py currently
use assertTrue and assertFalse instead of failIf and failUnless
fix a bit of a grammatical thinko in the all-new remove docstring


refs [#3097](https://github.com/twisted/twisted/issues/3097)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

Okay I like this now, I think I'll merge it.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set status to `closed`|
|-|-|

(In [23062]) Merge filepath-remove-symlink-3097

Author: cyli
Reviewer: therve, exarkun
Fixes [#3097](https://github.com/twisted/twisted/issues/3097)

Change FilePath.remove so that symlinks are only deleted by it, not the contents
of the directories they point to, and so that when a symlink which points to a
directory is encountered, an exception is not raised.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
